### DISCUSSION TITLE
So, this `heap` thing

### DISCUSSION THREAD

**Discussion by reinvantveer** [ricky26/rust-mega-drive#5] [Q&A]
Seeing how far I can stretch this - `alloc::vec::Vec` does not require the std lib and is immensely helpful in dynamic data wrangling. But, it requires something more than just `Vec::new()` I found out:

```
error: no global memory allocator found but one is required; link to std or add `#[global_allocator]` to a static item that implements the GlobalAlloc trait.

error: `#[alloc_error_handler]` function required, but not found.

note: Use `#![feature(default_alloc_error_handler)]` for a default error handler.
```
I did find [`pub unsafe fn heap() -> &'static mut [u8]`](https://github.com/ricky26/rust-mega-drive/blob/main/libs/megadrive-sys/src/lib.rs#L54) but this appears to just hand me a slice of memory available for dynamic allocation.  

So maybe this is beyond the capabilities here, but is there a chance to use, say, `Box` and `Vec` from the `alloc` compiler crate?

**Comment by reinvantveer** [comment]
I guess the `#![feature(default_alloc_error_handler)]` would be as easy to implement (as a dummy) as `#[panic_handler]` (maybe even in the same function???)

**Reply by reinvantveer** [reply]
Hm no looking at https://doc.rust-lang.org/std/alloc/trait.GlobalAlloc.html#example this requires something else. But the example there makes heavy use of `std`lib stuff

**Comment by ricky26** [comment]
The `heap()` function returns all the RAM which isn't statically allocated. There's probably a `GlobalAlloc` implementation which takes a fixed slice as the heap. A quick Google suggests something like https://crates.io/crates/wee_alloc could work.

`#![feature(default_alloc_error_handler)]` is an alternative to implementing `#[alloc_error_handler]` IIUC.

**Reply by ricky26** [reply]
Actually, https://github.com/phil-opp/linked-list-allocator looks like it could work.

**Reply by reinvantveer** [reply]
Alright sounds cool, I'll see if I can wrap my head around it

**Reply by reinvantveer** [reply]
Had a few goes at it, but this crate (although it looks ideally suited) is very hard to install. Versions of v0.7.0 and earlier aren't compatible with Rust nightly 1.53 due to
```
   Compiling linked_list_allocator v0.7.0
error[E0432]: unresolved imports `alloc::alloc::AllocErr`, `alloc::alloc::AllocRef`
  --> /root/.cargo/registry/src/github.com-1ecc6299db9ec823/linked_list_allocator-0.7.0/src/lib.rs:14:20
   |
14 | use alloc::alloc::{AllocErr, AllocRef, Layout};
   |                    ^^^^^^^^  ^^^^^^^^ no `AllocRef` in `alloc`
   |                    |
   |                    no `AllocErr` in `alloc`
   |                    help: a similar name exists in the module: `AllocError`
```

The more recent ones have other ailments. I think the closest comes 0.8.11 
```
linked_list_allocator = { version = "0.8", default-features = false, features = ["use_spin_nightly"] }
```
with dependency tree
```
├── linked_list_allocator v0.8.11
│   └── spinning_top v0.1.1
│       └── lock_api v0.3.4
│           └── scopeguard v1.1.0
```
which fails to compile with
```
error[E0432]: unresolved import `core::sync::atomic::AtomicUsize`
  --> /root/.cargo/registry/src/github.com-1ecc6299db9ec823/lock_api-0.3.4/src/remutex.rs:19:20
   |
19 |     sync::atomic::{AtomicUsize, Ordering},
   |                    ^^^^^^^^^^^ no `AtomicUsize` in `sync::atomic`

```
Seeing https://github.com/Amanieu/parking_lot/issues/277, I'm having a wild guess here that rust-m68k doesn't support atomics?

**Reply by reinvantveer** [reply]
Ah I see now that this error is linked to the LockedHeap implementation from the docs quickstart. I may have to dig deeper.

**Reply by reinvantveer** [reply]
Thanks to Philipp Opperman's replies on my attempts at `LockedHeap` in https://github.com/phil-opp/linked-list-allocator/issues/52 I'm now somewhat wiser. There's no `AtomicUsize` available on the rust_m68k, which makes things a lot harder to have a simple `LockedHeap` import and initialization for the rust-megadrive. Probably the usage of AtomicUsize makes the LockedHeap eligible for multi-threaded processing but honestly this seems a little bit too ambitious for a processor as simple as the m68k.

Unless it's trivially simple to add atomic instructions to the m68k target, but I imagine it isn't.

**Reply by ricky26** [reply]
Atomic instructions on the 68k are _largely_ useless since it's single core and there's no preemption. :smile: I guess it's just hard-coded into that library... We probably need a way dumber implementation.

(Actually, I guess the interrupts do preempt the rest of the code - so you'd have to be careful with that.)

**Reply by reinvantveer** [reply]
https://crates.io/crates/wee_alloc is going to be tough I think, because it relies on libc. Just tried it and I doubt this will resolve into a working allocator

**Reply by reinvantveer** [reply]
I got quite far just by cherry-picking stuff from linked_list_allocator. But here's something that confuses me.
This is the https://doc.rust-lang.org/core/alloc/trait.GlobalAlloc.html docs on GlobalAlloc. I believe that currently, we have no way around GlobalAlloc to implement allocation in a way that is transparent to use your garden variety `alloc::vec::Vec`

But how do you keep track of state on heap allocation if the GlobalAlloc trait takes an *immutable* reference?
![image](https://user-images.githubusercontent.com/9550733/123854360-e6d69180-d91e-11eb-9f9c-905ca83fe6dc.png)

**Reply by reinvantveer** [reply]
I'll open a PR, maybe this will clarify things a bit: https://github.com/ricky26/rust-mega-drive/pull/7

**Reply by ricky26** [reply]
The global allocator has to take immutable references since it can be used concurrently between threads. Generally you'd mitigate it with something like `Mutex` but alas, we don't got one of those here (yet). :smile: 

I'll try and find some time to look through your PR soon.

**Reply by reinvantveer** [reply]
:+1: cool

**Reply by reinvantveer** [reply]
I'm trying to wrap it in [`no_mutex`](https://crates.io/crates/no_mutex)

**Reply by reinvantveer** [reply]
Updated the PR description with a small todo-list, let me know if something else comes to mind.

**Reply by parasyte** [reply]
This is an interesting thread because I have been looking at options for building a `GlobalAlloc` implementation on Nintendo 64. Some day I want to use `rust-mega-drive` to bootstrap an SDK for Atari Jaguar, and I'll want an allocator on that platform as well.

It's trivial to implement a toy bump allocator for testing `alloc::collections` on your target. See [this playground example](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=6b4d93ca1e8190ceea9a0b5ae04f16fb) for a minimal allocator setup.

For the `GlobalAlloc` trait, the `alloc` and `dealloc` methods can be implemented with any type that has interior mutability. This is usually a `Mutex`, but you can also use the `UnsafeCell` family of types, which is what the playground example does.

Bump allocators are not an ideal allocator design for robust software. And this is why I've been exploring available options. I do have atomics on N64, so it's easy enough to use crates like `spin` to provide a `Mutex`. In your case with the 68000, you need to implement `Sync` with a platform-specific interrupt inhibit method. In short, you can use either `wee_alloc` or `linked-list-allocator`. You just have to disable default features and implement the `GlobalAlloc` trait yourself.

**Reply by reinvantveer** [reply]
Hi @parasyte thanks for your sharing your thoughts on this! I also saw your suggestions on #7 and I'll look into them soon. `UnsafeCell` sounds like a better fit - I doubt anyone would benefit from multiple threads on this platform. `spin` and `linked-list-allocator` all require something out of the box that is unavailable for m68k targets thus far, usually something like atomics, a different Rust nightly version or full-blown 32-bit multiplication (tracking issue #4).

The Atari Jaguar looks awesome. The whole Tom & Jerry ASIC situation sounds like quite a challenge! I've been doing some stuff on the Docker end that may be of interest: I recently published a [generic m68k toolchain image](https://quay.io/repository/reinvantveer/rust-m68k?tab=tags) based on @ricky26 's fixes in this [Dockerfile](https://github.com/reinvantveer/rust-mega-drive/blob/workflows/Dockerfile.toolchain) in a working branch that may be helpful and can help you save a couple of hours compiling llvm and rust.

**Reply by parasyte** [reply]
That docker image is excellent! Thank you for that, it should really reduce the entry barrier for Rust on m68k platforms!

The lack of intrinsics for #4 is too bad, but understandable. I'm surprised that `linked-list-allocator` needs 32-bit multiplies, but what can you do?

**Reply by reinvantveer** [reply]
Yeah the 32-bit multis are somewhere far under the hood of either rustc or LLVM, I havent seen any direct trace of them so far. On a good note: I'm pretty proud that I managed to implement the alloc crate with the unsafe cell implementation, to be honest I had quite the struggle with the borrow checker and I don't seem to get rid of `#![feature(const_mut_refs)]`.
